insonmnia/miner/builder.go

```diff
@@ -223,15 +229,71 @@ func (b *MinerBuilder) getPublicIPs() (pubIPs []string, err error) {
 	return nil, errors.New("failed to get public IPs")
 }
 
+func (b *MinerBuilder) getHubConnectionInfo() (string, common.Address, error) {
+	var hubEndpoint = b.cfg.HubEndpoint()
+	if strings.Contains(hubEndpoint, "@") {
+		return util.ParseEndpoint(b.cfg.HubEndpoint())
+	} else {
+		resolved, err := b.locatorClient.Resolve(b.ctx, &pb.ResolveRequest{EthAddr: hubEndpoint})
+		if err != nil {
+			return "", common.Address{}, fmt.Errorf(
+				"failed to resolve hub addr from %s: %s", hubEndpoint, err)
+		}
+
+		log.G(b.ctx).Info("resolved hub endpoints", zap.Any("endpoints", resolved.IpAddr))
+
+		for _, addr := range resolved.IpAddr {
+			addr = strings.Replace(addr, "10001", "10002", 1)
+
+			log.G(b.ctx).Debug("trying hub endpoint", zap.Any("endpoint", addr))
+
+			dialer := net.Dialer{DualStack: true, Timeout: time.Second}
+			testCC, err := dialer.DialContext(b.ctx, "tcp", addr)
+			if err != nil {
+				log.G(b.ctx).Debug(
+					"hub endpoint is unreachable", zap.Any("endpoint", addr),
+					zap.Any("error", err))
+			} else {
+				testCC.Close()
+				hubSockaddr, hubEthAddr, err := util.ParseEndpoint(hubEndpoint + "@" + addr)
+				if err == nil {
```

Comment:
what error will be returned if err is not nil?

Reply:
If `err` is not `nil` (which is sort of impossible in this case, but it's just a safety check) we move on to the next resolved address, and if none of the resolved addresses worked, we return an error.

Reply:
Oh, I get it, I'll make it return the most relevant error.